File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Easy/2373. Largest Local Values in a Matrix/C++ solution/2373. Largest Local Values in a Matrix.cpp
class Solution {
public:
    vector<vector<int>> largestLocal(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<vector<int>> maxLocal(n - 2, vector<int>(n - 2));

        for (int i = 0; i < n - 2; i++)
        {
            for (int j = 0; j < n - 2; j++)
            {
                int maxval = 0; // since the constraint told us that the valuse are greater then 1
                for (int row = i; row < i + 3; row++)
                {
                    for (int col = j; col < j + 3; col++)
                    {
                         maxval = max(maxval, grid[row][col]);
                    }
                }
                maxLocal[i][j] = maxval;
            }
        }
        return maxLocal;
    }
};

// # Or we can calculate the maxval by precalculating the max using the following method 
// # It is important to note that the following method is not scalable but the above method is, we can change 3 to any value say k which can be any value of a contigious K x K matrix

//  # Calculate the maximum value for the 3x3 grid starting at (i, j) This mehtod is not scalable
//                 # maxval = max(
//                 #     grid[i][j], grid[i][j+1], grid[i][j+2],
//                 #     grid[i+1][j], grid[i+1][j+1], grid[i+1][j+2],
//                 #     grid[i+2][j], grid[i+2][j+1], grid[i+2][j+2]
//                 # )

File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Easy/2373. Largest Local Values in a Matrix/Python soultion/2373. Largest Local Values in a Matrix.py
class Solution:
    def largestLocal(self, grid: List[List[int]]) -> List[List[int]]:
        n = len(grid)
        maxLocal = [[0] * (n - 2) for _ in range(n - 2)]

        for i in range(n - 2): # since the grid is a square matrix
            for j in range(n - 2):
                maxval = 0
                for row in range(i, i+3):
                    for col in range(j, j+3):
                        maxval = max(maxval, grid[row][col])
            
                maxLocal[i][j] = maxval
        
        return maxLocal

# Or we can calculate the maxval by precalculating the max using the following method 
# It is important to note that the following method is not scalable but the above method is, we can change 3 to any value say k which can be any value of a contigious K x K matrix

 # Calculate the maximum value for the 3x3 grid starting at (i, j) This mehtod is not scalable
                # maxval = max(
                #     grid[i][j], grid[i][j+1], grid[i][j+2],
                #     grid[i+1][j], grid[i+1][j+1], grid[i+1][j+2],
                #     grid[i+2][j], grid[i+2][j+1], grid[i+2][j+2]
                # )